apps/devhub/utils.py

```diff
@@ -84,3 +97,203 @@ def escape_validation(validation):
+                     .exclude(~Q(platform=self.file.platform) &
+                              Q(version__files__platform=self.file.platform))
+                     # Files which are not for the either the same platform or
+                     # for all platforms, but have other versions in the same
+                     # version which are.
+                     .exclude(~Q(platform__in=(self.file.platform,
+                                               amo.PLATFORM_ALL.id)) &
+                              Q(version__files__platform=amo.PLATFORM_ALL.id)))
+
+        for file in files:
+            # Only accept versions which come before the one we're validating.
+            if Version(file.version.version) < version:
+                return file
+
+
+class ValidationComparator(object):
+    """Compares the validation results from an older version with a version
+    currently being checked, and annotates results based on which messages
+    may be ignored for the purposes of automated signing."""
+
+    def __init__(self, validation):
+        self.validation = validation
+
+        self.messages = {self.message_key(msg): msg
+                         for msg in validation['messages']
+                         if 'context' in msg}
+
+    @classmethod
+    def message_key(cls, message):
+        """Returns a hashable key for a message based on properties
+        required when searching for a match."""
+
+        # No context, message is not matchable.
+        if 'context' not in message:
+            return None
+
+        return (tuple(message['id']),
+                tuple(message['context']),
+                message['file'],
+                message.get('signing_severity'),
+                ('context_data' in message and
+                 tuple(message['context_data'].iteritems())))
+
+    @classmethod
+    def match_messages(cls, message1, message2):
+        """Returns true if the two given messages match. The two messages
+        are assumed to have identical `message_key` values."""
+
+        # We'll eventually have to deal with issues like line number
+        # matching, with some kind of slop factor. For now, just rely
+        # on the basic heuristic of file name and context data.
+        return True
+
+    def find_matching_message(self, message):
+        """Finds a matching message in the saved validation results,
+        based on the return value of `message_key` and `match_messages`,
+        and return it. If not matching message exists, return None."""
+
+        msg = self.messages.get(self.message_key(message))
+        if msg and self.match_messages(msg, message):
+            return msg
+
+    @classmethod
+    def is_ignorable(cls, message):
+        """Returns true if a message may be ignored when a matching method
+        appears in past results."""
+
+        low_severity = message['signing_severity'] in ('trivial', 'low')
+        return message.get('ignore_duplicates', low_severity)
```

Comment:
It doesn't exist yet. It's the flag that will be added when an editor decides if a message should or should not be ignored, despite the `signing_severity`.